src/src/Commands/GPIO.cpp

```diff
@@ -0,0 +1,392 @@
+    const uint32_t key = createKey(pluginID,event->Par1);
+    // WARNING: operator [] creates an entry in the map if key does not exist
+    // So the next command should be part of each command:
+    byte mode;
+    int8_t state;
+
+    if (existPortStatus(key))
+    {
+      mode=globalMapPortStatus.at(key).mode;
+      state=globalMapPortStatus.at(key).state;
+    } else {
+      GPIO_Read(pluginID, event->Par1, &state);
+      mode = (state==-1)?PIN_MODE_OFFLINE:PIN_MODE_OUTPUT;
+    }
+
+    switch (mode) {
+      case PIN_MODE_OUTPUT:
+      case PIN_MODE_UNDEFINED:
+        {
+          createAndSetPortStatus_Mode_State(key,PIN_MODE_OUTPUT,!state);
+          GPIO_Write(pluginID, event->Par1, !state);
+
+          String log = logPrefix + String(F(" toggle: port#")) + String(event->Par1) + String(F(": set to ")) + String(!state);
+          addLog(LOG_LEVEL_ERROR, log);
+      	  SendStatusOnlyIfNeeded(event->Source, SEARCH_PIN_STATE, key, log, 0);
+
+          return return_command_success();
+        }
+        break;
+      case PIN_MODE_OFFLINE:
+        logErrorGpioOffline(logPrefix,event->Par1);
+        return return_command_failed();
+        break;
+      default:
+        logErrorGpioNotOutput(logPrefix,event->Par1);
+        return return_command_failed();
+        break;
+    }
+  } else {
+    logErrorGpioOutOfRange(logPrefix,event->Par1);
+    return return_command_failed();
+  }
+}
+
+String Command_GPIO(struct EventStruct *event, const char* Line)
+{
+  String logPrefix = new char;
+  byte pluginID;
+  //Line[0]='g':gpio; ='p':pcfgpio; ='m':mcpgpio
+  bool success = getPluginIDAndPrefix(Line[0], &pluginID, &logPrefix);
+  if (success && checkValidPortRange(pluginID, event->Par1))
+  {
+	  int8_t state=0;
+	  byte mode;
+
+	  if (event->Par2 == 2) { //INPUT
+		  mode = PIN_MODE_INPUT_PULLUP;
+      switch (pluginID) {
+        case PLUGIN_GPIO:
+          setGPIOPullupMode(event->Par1);
+          state = GPIO_Read_Switch_State(event->Par1, PIN_MODE_INPUT_PULLUP);
+          break;
+        case PLUGIN_MCP:
+        case PLUGIN_PCF:
+          // PCF8574/MCP specific: only can read 0/low state, so we must send 1
+          state = 1;
+          break;
+		  }
+    } else { // OUTPUT
+      mode=PIN_MODE_OUTPUT;
+      state=(event->Par2==0)?0:1;
+    }
+
+    const uint32_t key = createKey(pluginID,event->Par1);
+
+    if (globalMapPortStatus[key].mode != PIN_MODE_OFFLINE)
+    {
+      int8_t currentState;
+      GPIO_Read(pluginID, event->Par1, &currentState);
+      if (currentState==-1) {
+        mode=PIN_MODE_OFFLINE;
+        state = -1;
+      }
+
+      createAndSetPortStatus_Mode_State(key,mode,state);
+
+      switch (pluginID) {
+        case PLUGIN_GPIO:
+          if (mode==PIN_MODE_OUTPUT) GPIO_Internal_Write(event->Par1, state);
+  				break;
+  			case PLUGIN_MCP:
+  				GPIO_MCP_Write(event->Par1, state);
+  				break;
+  			case PLUGIN_PCF:
+  				GPIO_PCF_Write(event->Par1, state);
+  				break;
+      }
```

Comment:
Why not use your new function `GPIO_Write` here?

Reply:
Will do. It's not fully optimized yet...

Reply:
done

Reply:
done